### DISCUSSION TITLE
HighDPI for macOs - how to enable or disable it?

### DISCUSSION THREAD

**Discussion by ramboza** [raysan5/raylib#2999] [General]
Hello guys,

Is any verified guide how to enable HighDPI for Mac available? 
i tired  SetConfigFlags(FLAG_WINDOW_HIGHDPI)  and  "NSHighResolutionCapable" with Info.plist - is not working. 

My macOs is Ventura and raylib 4.5.0

Thanks,
Jev.

**Comment by staatzstreich** [comment]
Not a guide...  but perhaps it helps a little.

As far as I can say, with my own research on the net and testing on a 4K Monitor owned by an co worker.... I used:

```
const int screenWidth = 800;
const int screenHeight = 450;
const int gameScreenWidth = 1920;
const int gameScreenHeight = 1080;
```

I use a virtual game screen. there is an raylib - example for this at: "raylib [core] example - window scale letterbox (and virtual mouse)"
 

for my simple patience card game and the window setting is

```
SetConfigFlags(FLAG_WINDOW_RESIZABLE | FLAG_VSYNC_HINT | FLAG_WINDOW_HIGHDPI);
InitWindow(screenWidth, screenHeight, "PatienceMax");
SetWindowMinSize(320, 240);
```

I can't remember exactly why but afaik the graphics should named with fx. "my_graphic@2x.png" and should twice bigger

hope that helps a little.

Best,
Michael

**Comment by ramboza** [comment]
Hello Michael, 

Thank you for reply!!  of course i tried to use FLAG_WINDOW_HIGHDPI flag ...  also with/without it result the same.

here is my simple example: 
```c++

#include "raylib.h"

int main(int argc, const char * argv[]) {

	const int screenWidth = 400;
	const int screenHeight = 400;

	SetConfigFlags(FLAG_WINDOW_RESIZABLE | FLAG_VSYNC_HINT | FLAG_WINDOW_HIGHDPI);
	InitWindow(screenWidth, screenHeight, "dpi-test");
	SetTargetFPS(60);

	Image image = LoadImage("/path/to/PrintDPI.png");
	Texture2D texture = LoadTextureFromImage(image);
	UnloadImage(image);

	while (!WindowShouldClose()) {
		if (IsKeyPressed(KEY_ESCAPE)) break;

		BeginDrawing();
			ClearBackground(GRAY);
			DrawText("THIS IS TEST 1234 [] -_/\\!", 100, 50, 12, BLACK);

			DrawTexture(texture, 100, 100, WHITE);
		EndDrawing();
	}

	CloseWindow();
	return 0;
}

```

and here is result 
 
<img width="1142" alt="Screenshot 2023-04-07 at 00 45 27" src="https://user-images.githubusercontent.com/2714575/230508462-2679e164-0460-4172-82d0-c88a798c64b9.png">


By the way - using windows 11 it works  (I can see smaller window and high resolution picture) 

Also  i tried to put binary inside of "Some.app" with Contents/Info.plist and
...
   <key>NSHighResolutionCapable</key>
        <true/>
...
Result the same :(

**Comment by staatzstreich** [comment]
As far as I know... Windows is different from MacOS....

From Web search (I am a no native English speaker):

HiDPI mode means that the display stays at a certain resolution but draws everything scaled twice.

E.g. a 3840x2160 (4K) display can work as a huuuuge display with the screen real estate of four (normal) FullHD displays.
Or in HiDPI mode the screen real estate will be equal to a FullHD display but displayed with four times as many pixels (much sharper text, windows, icons).
The screen of your MacBook runs at 2880x1800 HiDPI — the screen real estate is equal to any other 1440x900 laptop, but is sharper. Apple calls this "retina display".

A "scaled resolution" is a term Apple uses for something that's really just feeding a non-native resolution to the display. For example, I'm running my 13" at 1440x900, scaled. The display itself is 2560x1600.
This is what happens:
- the screen is composed (window sizes, positions) as if it was a 1440x900 screen
- it's a retina screen (HiDPI), so everything is rendered x2, i.e. at 2880x1800
- the 2880x1800 result is displayed on a 2560x1600 screen
Note that the resolutions don't match each other, so each physical pixel displays the result of interpolation between two virtual pixels. This causes the image to be sliiiightly blurrier, which is a hit I'm willing to accept for more screen estate. This is the whole magic behind "scaled resolution". (There's no magic)

Windows handles scaling differently - window and font sizes are actually changing by a percentage. This produces clear and sharp images but is much harder to implement and can cause weird compatibility problems.


That describes what I mean... on Mac your **image size** must be greater not the dpi itself.  I add some screens from my game. hope that clarifies what I meant .

<img width="803" alt="Bildschirmfoto 2023-04-07 um 04 35 18" src="https://user-images.githubusercontent.com/2380482/230530680-22744155-2c26-4321-ae7d-bf1919e43e49.png">

The Image of a card, used in the game is only 113x157 pixel. This way the hole game screen can scale... see the next 2 screen shots

<img width="824" alt="Bildschirmfoto 2023-04-07 um 04 35 44" src="https://user-images.githubusercontent.com/2380482/230531037-ebdd36c9-1810-4036-8dee-f12bfaed162b.png">
<img width="1325" alt="Bildschirmfoto 2023-04-07 um 04 36 02" src="https://user-images.githubusercontent.com/2380482/230531056-091f1789-6e1d-4dfb-8609-abd4101bc8da.png">

and that way it is also looks good on a 4K Display connected to a Mac.


I have no idea if this works on windows ... I stopped using Windows at Windows98 and never looked back :)


best,
Michael

**Comment by ramboza** [comment]
Hello Michael,

For sure I know about highDPI on win/macOS and anyway Thank you for explanation! 
Let me show one really simple example to make sense .... So, SDL library, parameter `SDL_WINDOW_ALLOW_HIGHDPI` will affect display. 
 
<img width="799" alt="Screenshot 2023-04-10 at 12 40 52" src="https://user-images.githubusercontent.com/2714575/230888051-21b7de36-a5fa-4556-b6fd-8a63f45963df.png">

Without this parameter image will not fit pixel->pixel on the screen (which is HighDPI).
<img width="723" alt="Screenshot 2023-04-10 at 12 41 22" src="https://user-images.githubusercontent.com/2714575/230888057-48bfbed5-8dd4-4b00-bd8e-d611661e5732.png">

Question is:
Raylib has a similar flag `FLAG_WINDOW_HIGHDPI`, but usage of this flag does not affect how loaded image will displayed (with/without will use LowDPI way). 
It looks like a HighDPI hint is not working. Maybe my approach should be different ... want to know how to make the same as SDL does. 

Under the hood of raylib, GLFW in use and maybe something wrong with it? 
https://www.glfw.org/docs/latest/group__window.html#gaf5d31de9c19c4f994facea64d2b3106c
 

Thanks,
Jev.

**Comment by staatzstreich** [comment]
Hello Jev… i am sorry but i am not an expert on this topic…
May be you will Look here:
https://github.com/raysan5/raylib/issues/1086
or wait for getting answer from someone else 

Best,
Michael

**Reply by ramboza** [reply]
Thanks Michael,

I figured it out! A little weird, but it works!
Strange, because function `DrawTexture()` will not support High DPI, `DrawTexturePro` will work.

GetWindowScaleDPI() should be used to get actual 'dpi' scale.


```c++
#include "raylib.h"
#include "GLFW/glfw3.h"
#include <iostream>
 
int main(int argc, const char * argv[]) {

	const int screenWidth = 400;
	const int screenHeight = 400;

	SetConfigFlags(FLAG_MSAA_4X_HINT | FLAG_WINDOW_ALWAYS_RUN | FLAG_WINDOW_HIGHDPI);
	InitWindow(screenWidth, screenHeight, "dpi-test");
	SetTargetFPS(60);

	Vector2 scale = GetWindowScaleDPI();
	Image image = LoadImage("/Users/jenya/Desktop/PrintDPI.png");
	Texture2D texture = LoadTextureFromImage(image);
	UnloadImage(image);

	char out[100];
	while (!WindowShouldClose()) {
		if (IsKeyPressed(KEY_ESCAPE)) break;
		BeginDrawing();
			ClearBackground(GRAY);

			// DrawTexture(texture, 0, 0, WHITE); << not supported HighDPI
			
			// dst rectangle should use scale 1.0 or 2.0 returned by GetWindowScaleDPI()
			Rectangle src {0, 0, (float)texture.width, (float)texture.height};
			Rectangle dst {0, 0, (float)texture.width / scale.x, (float)texture.width / scale.y};
			
			DrawTexturePro(texture, src, dst, (Vector2) {0, 0}, 0, WHITE);
		EndDrawing();
	}

	CloseWindow();
	return 0;
}

```

**Reply by staatzstreich** [reply]
Hi Jevgenijs,

thanks for sharing.... 

Best,
Michael
